bootstrap/src/main/java/jenkins/bootstrap/Plugin.java

```diff
@@ -0,0 +1,188 @@
+ * into the classpath, the bootstrap logic needs to inspect plugins lightly.
+ * Jenkins will later do further full treatment of plugins.
+ *
+ * <p>
+ * Some of the capabilities of the full plugin manager is missing from this code. For example,
+ * the bootstrap doesn't handle {@code *.jpl} file used during the development.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+class Plugin {
+    /**
+     * foo.jpi file
+     */
+    private final File archive;
+    /**
+     * Directory to explode the plugin.
+     */
+    private final File destDir;
+
+    Plugin(File archive) {
+        this.archive = archive;
+        this.destDir = getDestDir();
+    }
+
+    void process(DependenciesTxt core, List<URL> jars, List<DependenciesTxt> overrides) throws IOException {
+        if (!isCoreComponentOverride() || isDisabled())
+            return;
+
+        explode();
+
+        File dependencies = new File(destDir,"META-INF/dependencies.txt");
+        try (InputStream i = new FileInputStream(dependencies)) {
+            DependenciesTxt deps = new DependenciesTxt(i);
+            for (Dependency d : deps.dependencies) {
+                if (core.hasNewerThan(d)) {
+                    // core has a newer version. abort.
+                    LOGGER.warning("Skipping core override plugin "+archive+" because core contains a newer version of "+d.ga);
+                    return;
+                }
+            }
+            overrides.add(deps);
+        }
+
+
+        File lib = new File(destDir, "WEB-INF/lib");
+        LOGGER.info("Loading core overrides from "+lib);
+        File[] files = lib.listFiles();
+        if (files!=null) {
+            for (File jar : files) {
+                if (!jar.getName().endsWith(".jar"))
+                    continue;
+                // TODO: version check
+                jars.add(jar.toURI().toURL());
+            }
+        }
+    }
+
+    /**
+     * Gets the directory name to explode the jpi file into.
+     */
+    private File getDestDir() {
+        String name = archive.getName();
+        int idx = name.lastIndexOf('.');
+        if (idx<0)  throw new IllegalArgumentException("No extension in file name: "+archive);
+        return new File(archive.getParentFile(), name.substring(0,idx));
+    }
+
+    /**
+     * Returns true if this plugin overrides core components.
+     */
+    private boolean isCoreComponentOverride() throws IOException {
+        Manifest mf = loadManifest();
+        return "true".equals(mf.getMainAttributes().getValue("Core-Component"));
+    }
+
+    public boolean isDisabled() {
+        File disableFile = new File(archive.getPath() + ".disabled");
+        return disableFile.exists();
+    }
+
+    /**
+     * Loads jar file manifest from the archive.
+     */
+    private Manifest loadManifest() throws IOException {
+        try (JarFile jar = new JarFile(archive)) {
+            ZipEntry e = jar.getEntry("META-INF/MANIFEST.MF");
```

Comment:
We need to  ensure there is no File descriptor leaks in this logic. There were some recent issues in other places